src/storage.rs

```diff
@@ -90,207 +242,180 @@ impl<T: 'static> RingBufferStorage<T> {
+    pub fn single_write(&mut self, element: T) {
+        self.ensure_additional(1);
+        println!(
+            "Single write, last index: {:?}, putting at {}",
+            self.last_index,
+            self.last_index + 1
+        );
+        unsafe {
+            self.data.put(self.last_index + 1, element);
+        }
+        self.last_index += 1;
+        self.generation += Wrapping(1);
+    }
+
+    /// Create a new reader id for this ring buffer.
     pub fn new_reader_id(&mut self) -> ReaderId<T> {
-        let alive;
-        // Attempt to re-use positions from dropped readers.
-        let new_id = self.reader_internal
-            .iter()
-            .map(|ref internal| unsafe { &*internal.get() })
-            .position(|internal| !internal.alive.load(Ordering::Relaxed));
-        match new_id {
-            Some(new_id) => {
-                let reader = unsafe { &mut *self.reader_internal[new_id].get() };
-                reader.written = self.written;
-                reader.index = self.write_index;
-                reader.alive.store(true, Ordering::Relaxed);
-                alive = reader.alive.clone();
-            }
-            None => {
-                alive = Arc::new(AtomicBool::new(true));
-                self.reader_internal.push(UnsafeCell::new(InternalReaderId {
-                    written: self.written,
-                    index: self.write_index,
-                    alive: alive.clone(),
-                }));
-            }
+        let meta = self.meta.get_mut();
+        let last_index = self.last_index.index;
+        let generation = self.generation.0;
+        let id = meta.free.pop().unwrap_or_else(|| {
+            let id = meta.readers.len();
+            meta.readers.push(Reader {
+                generation,
+                last_index,
+            });
+
+            id
+        });
+
+        ReaderId {
+            id,
+            marker: PhantomData,
         }
-        ReaderId::new(
-            new_id.unwrap_or(self.reader_internal.len() - 1),
-            self.buffer_id,
-            alive,
-        )
     }
 
-    /// Read data from the ringbuffer, starting where the last read ended, and up to where the last
-    /// data was written.
+    /// Read data from the ring buffer, starting where the last read ended, and up to where the last
+    /// element was written.
     pub fn read(&self, reader_id: &mut ReaderId<T>) -> StorageIterator<T> {
-        assert!(
-            reader_id.buffer_id == self.buffer_id,
-            "ReaderID used with an event buffer it's not registered to.  Not permitted!"
-        );
-        let written = unsafe { (*self.reader_internal[reader_id.reader_id].get()).written };
-        let num_written = if self.written < written {
-            self.written + (self.reset_written - written)
-        } else {
-            self.written - written
+        let (last_read_index, gen) = {
+            let mut meta = self.meta.lock();
```

Comment:
Is this necessary, since we can guarantee noone's writing when a read occurs?

Reply:
Yeah well some other reader could access the meta data as well (and write to it). I'm aware we can use `UnsafeCell`, but that would require to add more checks (like does this id belong to this buffer).